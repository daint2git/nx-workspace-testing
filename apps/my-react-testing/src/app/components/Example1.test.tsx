import { DataGrid } from '@mui/x-data-grid';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { mocked } from 'ts-jest/utils';
import Example1, { rows } from './Example1';

jest.mock('@mui/x-data-grid', () => ({
  ...jest.requireActual('@mui/x-data-grid'),
  DataGrid: jest.fn(() => <div>Mocked DataGrid</div>),
}));

// use jest.mocked instead of https://github.com/facebook/jest/blob/main/CHANGELOG.md#2744
const mockedDataGrid = mocked(DataGrid);

describe('Example1', () => {
  beforeEach(() => {
    mockedDataGrid.mockClear();
  });

  it('renders MUI DataGrid with columns and rows', () => {
    const onMoneyMockFn = jest.fn();

    render(<Example1 onMoney={onMoneyMockFn} />);

    userEvent.click(screen.getByRole('button', { name: 'Give me 33 dollars' }));

    expect(onMoneyMockFn).toBeCalledTimes(1);
    expect(onMoneyMockFn).toBeCalledWith(33);
  });

  it('renders DataGrid passing the expected props', () => {
    const onMoneyMockFn = jest.fn();

    const { container } = render(<Example1 onMoney={onMoneyMockFn} />);

    expect(mockedDataGrid).toBeCalledTimes(1);

    // console.log('mockedDataGrid.mock.calls', mockedDataGrid.mock.calls);
    // result:
    // mockedDataGrid.mock.calls [
    //   [
    //     {
    //       rows: [Array],
    //       columns: [Array],
    //       pageSize: 5,
    //       rowsPerPageOptions: [Array],
    //       checkboxSelection: true
    //     },
    //     {}
    //   ]
    // ]

    expect(mockedDataGrid).toBeCalledWith(
      {
        rows: rows,
        columns: [
          expect.objectContaining({ field: 'id' }),
          expect.objectContaining({ field: 'firstName' }),
          expect.objectContaining({ field: 'lastName' }),
          expect.objectContaining({ field: 'age' }),
          expect.objectContaining({ field: 'fullName' }),
        ],
        pageSize: 5,
        rowsPerPageOptions: [5],
        checkboxSelection: true,
      },
      {}
    );

    expect(container).toMatchSnapshot();
  });
});
