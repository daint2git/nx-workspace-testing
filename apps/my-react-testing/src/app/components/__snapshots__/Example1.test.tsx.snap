// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Example1 renders DataGrid passing the expected props 1`] = `
<div>
  <div
    class="MuiBox-root css-0"
  >
    <button
      class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButtonBase-root  css-1e6y48t-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Give me 33 dollars
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
    <div
      style="height: 400px; width: 100%;"
    >
      <div>
        Mocked DataGrid
      </div>
    </div>
  </div>
</div>
`;
